meddlr/metrics/SSFD.py

```diff
@@ -0,0 +1,205 @@
+        https://openreview.net/forum?id=dgMvTzf6M_3
+    """
+
+    def __init__(
+        self,
+        mode: str = "grayscale",
+        channel_names: Sequence[str] = None,
+        reduction="none",
+        compute_on_step: bool = False,
+        dist_sync_on_step: bool = False,
+        process_group: bool = None,
+        dist_sync_fn: bool = None,
+    ):
+
+        super().__init__(
+            channel_names=channel_names,
+            units="",
+            reduction=reduction,
+            compute_on_step=compute_on_step,
+            dist_sync_on_step=dist_sync_on_step,
+            process_group=process_group,
+            dist_sync_fn=dist_sync_fn,
+        )
+
+        valid_modes = ("grayscale", "rgb")
+        if mode not in valid_modes:
+            raise ValueError(
+                f"Argument `mode` must be one \
+                            of {valid_modes}, but got {mode}."
+            )
+
+        self.mode = mode
+
+        path_manager = env.get_path_manager()
+        file_path = path_manager.get_local_path(SSFD_HUGGINGFACE_URL, force=False)
+
+        self.net = SSFD_Encoder(in_channels=1, out_channels=320)
+        self.net.load_state_dict(torch.load(file_path))
+        self.net.eval()
+
+    def func(self, preds, targets) -> torch.Tensor:
+
+        if self.mode == "grayscale":
+            loss_shape = (targets.shape[0], targets.shape[1])
+
+        elif self.mode == "rgb":
+            if targets.shape[1] != 3:
+                raise ValueError(
+                    f"Channel dimension must have size 3 for rgb mode,\
+                    but got tensor of shape {targets.shape}."
+                )
+
+            loss_shape = (targets.shape[0], 1)
+
+        preds = self.preprocess_ssfd(preds)
+        targets = self.preprocess_ssfd(targets)
+
+        features = {}
+
+        def get_features(name):
+            def hook(model, input, output):
+                features[name] = output.detach()
+
+            return hook
+
+        self.net.encoder3.conv_block.conv2.register_forward_hook(
+            get_features("encoder3.conv_block.conv2")
+        )  # set default to this but make flexible
+        _ = self.net(targets)
+        target_features = features["encoder3.conv_block.conv2"]
+
+        _ = self.net(preds)
+        pred_features = features["encoder3.conv_block.conv2"]
+
+        loss = torch.mean(mse(target_features, pred_features), dim=1)
+        loss = loss.view(loss_shape)
+
+        return loss
+
+    def preprocess_ssfd(self, img):
+        """Preprocess image for SSFD model input. Converts to a magnitude scan and
+        normalizes to zero-mean, unit variance.
+
+        Args:
+            img (torch.Tensor): Tensor to preprocess
+
+        Returns:
+            Preprocessed tensor
+        """
+
+        is_complex = cplx.is_complex(img) or cplx.is_complex_as_real(img)
+        abs_func = cplx.abs if is_complex else torch.abs
+        img = abs_func(img)
```

Comment:
if the image is already real-valued, why do we take the magnitude? Wouldn't that make negative values positive?